course7/homeworks/patrickkostjens/ls/ls.cc

```diff
@@ -1,13 +1,60 @@
 #include <boost/filesystem.hpp>
+#include <boost/program_options.hpp>
 #include <iostream>
 
+namespace fs = boost::filesystem;
+namespace po = boost::program_options;
+
+size_t get_size(fs::path path)
+{
+	if (!fs::is_directory(path)) return fs::file_size(path);
+
+	size_t size = 0;
+	for (const auto& it : fs::directory_iterator(path)) 
+		size += get_size(it);
+	return size;
+}
+
+void print_contents(fs::path path, bool extensive, bool tree, int level = 0)
+{
+	const auto& is_directory = fs::is_directory(path);
+
+	std::cout << std::string(level * 2, ' ') << path.filename().string();
+	if (is_directory) std::cout << "/";
+	if (extensive) std::cout << ": " << get_size(path);
+	std::cout << std::endl;
+
+	if (is_directory && tree)
+		for (const auto& it : fs::directory_iterator(path))
```

Comment:
So here you descend recursively down the tree _again_. This means that you will call ```fs::file_size``` many times on the same file. This is an IO operation, potentially very slow.

Can you try re-structure the program so that you only visit every file once?